tools/eslint-plugin/dist/rules/public-extension-exports.js

```diff
@@ -0,0 +1,222 @@
+        declaration.declarationList.declarations[0].symbol.escapedName :
+        declaration.symbol.escapedName;
+      const kind = getSyntaxKindFriendlyName(declaration.kind);
+      const csvString = createCsvString(name, kind);
+      fs.writeFileSync(apiFilePath, csvString, { flag: "a" });
+    }
+
+    function getFileName(parent) {
+      let currentParent = parent;
+      while (currentParent) {
+        if (currentParent.fileName !== undefined)
+          return currentParent.fileName;
+        currentParent = currentParent.parent;
+      }
+      return undefined;
+    }
+
+    function isLocalFile(declaration) {
+      if (declaration) {
+        const fileName = getFileName(declaration.parent);
+        if (fileName && typeof fileName === "string" && !fileName.includes("node_modules"))
+          return true;
+      }
+      return false;
+    }
+
+    function getParentSymbolName(declaration) {
+      if (declaration.parent && declaration.parent.symbol && !declaration.parent.symbol.escapedName.startsWith('"'))
+        return declaration.parent.symbol.escapedName;
+      return undefined;
+    }
+
+    function checkJsDoc(declaration, node) {
+      // Only check local elements, not consumed ones
+      if (!declaration || !declaration.jsDoc || !isLocalFile(declaration))
+        return undefined;
+
+      for (const jsDoc of declaration.jsDoc)
+        if (jsDoc.tags) {
+          let jsDocExtensionTag = jsDoc.tags.find(tag => tag.tagName.escapedText === extensionsTag);
+          let jsDocPreviewTag = jsDoc.tags.find(tag => tag.tagName.escapedText === previewTag);
+          // Has extension API tag
+          if (jsDocExtensionTag) {
+            addToApiList(declaration, jsDocPreviewTag);
+            // Does not have any of the required release tags
+            if (!jsDoc.tags.some(tag => releaseTags.includes(tag.tagName.escapedText))) {
+              let name;
+              if (declaration.kind === ts.SyntaxKind.Constructor)
+                name = declaration.parent.symbol.escapedName;
+              else {
+                name = declaration.symbol.escapedName;
+                const parentSymbol = getParentSymbolName(declaration);
+                if (parentSymbol)
+                  name = `${parentSymbol}.${name}`;
+              }
+
+              context.report({
+                node,
+                messageId: "forbidden",
+                data: {
+                  kind: getSyntaxKindFriendlyName(declaration.kind),
+                  name,
+                  releaseTags: releaseTags,
+                }
+              });
+            }
+          }
+        }
+    }
+
+    function checkWithParent(declaration, node) {
+      if (!declaration)
+        return;
+      checkJsDoc(declaration, node);
+      if (declaration.parent && [
+        ts.SyntaxKind.ClassDeclaration,
+        ts.SyntaxKind.EnumDeclaration,
+        ts.SyntaxKind.InterfaceDeclaration,
+        ts.SyntaxKind.ModuleDeclaration,
```

Comment:
I'm not sure why this code is here - I ran this code a few times and the if statement is never true. If the class has the proper extension tags, then the whole class (and therefore all methods in said class) are included in the API. If the class is inside a namespace, then if the namespace has the @extension tag, the class will be included, if the namespace doesn't the class won't be. I think the non-extension API tags work the same way? If @calebmshafer can confirm then I'll get rid of this code.

Reply:
I haven't run the rule myself, but I'm guessing this code is supposed to cover cases where a class has only the `@extension` tag and its parent namespace has `@public`. In my understanding, the class should then inherit the "public" status.

Reply:
No, I think it shouldn't fail because everything inside `Test`, including `TestClass`, is public unless stated otherwise. From [api-extractor documentation](https://api-extractor.com/pages/tsdoc/doc_comment_syntax/#release-tags):

> The release tag applies recursively to members of a container (e.g. class or interface). For example, if a class is marked as @beta, then all of its members automatically have this status; you DON’T need add the @beta tag to each member function. However, you could add @internal to a member function to give it a different release status.

Reply:
I'm just going by what I perceive as "standard" release tag rules, but @calebmshafer might have a better idea if we want to do something different in this case.

Reply:
I think this is a different case than api-extractor since it cares more about the granular level than exports would. In terms of exports, if a namespaces is exported, all of its contents will be exported, and it should be designed with that in mind. @johnnyd710  is going to look into potentially failing linting for the case where a namespace member has the @extension tag but the namespace does not

Reply:
After some testing, the following would actually cause core-extension to fail to compile. You can't export/import TestClass without Test. My next commit will be an ESLint rule to catch if something inside a namespace has an extensions tag but the namespace doesn't.
```ts
/** @public */
export namespace Test {
  /** @extensions */
  export class TestClass {
  }
}
```